src/gadgets/non_native_field/implementations/implementation_u16.rs

```diff
@@ -1032,14 +1035,106 @@ impl<F: SmallField, T: pairing::ff::PrimeField, const N: usize> CSAllocatable<F>
     }
 }
 
+impl<F: SmallField, T: pairing::ff::PrimeField, const N: usize> CSPlaceholder<F>
+    for NonNativeFieldOverU16<F, T, N>
+{
+    fn placeholder<CS: ConstraintSystem<F>>(cs: &mut CS) -> Self {
+        let variable = Variable::placeholder();
+
+        Self {
+            limbs: [variable; N],
+            non_zero_limbs: 0,
+            tracker: OverflowTracker { max_moduluses: 0 },
+            form: RepresentationForm::Normalized,
+            params: Arc::new(NonNativeFieldOverU16Params::placeholder(cs)),
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<F: SmallField, T: pairing::ff::PrimeField, const N: usize> CircuitVarLengthEncodable<F>
+    for NonNativeFieldOverU16<F, T, N>
+{
+    fn encoding_length(&self) -> usize {
+        N
+    }
+
+    fn encode_to_buffer<CS: ConstraintSystem<F>>(&self, cs: &mut CS, dst: &mut Vec<Variable>) {
+        dst.extend_from_slice(self.limbs.as_slice())
+    }
+}
+
 // We need this to ensure no conflicting implementations without negative impls
 
-#[derive(Derivative)]
-#[derivative(Clone, Copy, Debug, Hash)]
+#[derive(Derivative, Serialize, PartialEq)]
```

Comment:
what was the reason for adding serialize functionality to this type?

Reply:
In circuits for `ec_pairing` we use internal FSM state with `Fq12` as intermediate value element, thus it was needed to add it here because otherwise it wouldn't get compiled